### PR TITLE
Support copying an image without a bmap file

### PR DESCRIPTION
**Pull request by Razaloc**
--nobmap or -n flag allows copying without a bmap file.
Closes: #39

Signed-off-by: Rafael Garcia Ruiz <rafael.garcia@collabora.com>

### CODE REVIEW

**Comment by Razaloc** [comment]
Things to consider:
- Avoid code repetition with match statements or repeat code at new copy function at main, as happened with the download option.  
- Since can't obtain image_size from bmap I've made the setup input to return the file size but we need to accommodate that elsewhere. Because is not considering compression.
- New lib copy_nobmap function.

**Comment by obbardc** [comment]
> Things to consider:
> 
> * Avoid code repetition with match statements or repeat code at new copy function at main, as happened with the download option.
> * Since can't obtain image_size from bmap I've made the setup input to return the file size but we need to accommodate that elsewhere. Because is not considering compression.

bmap-tools handles that by just showing a spinner and not the full progress bar. But if the file is compressed, and we're copying the whole uncompressed image, can we not just get the size of the uncompressed file from the decompression library ?

> * New lib copy_nobmap function.

**Comment by Razaloc** [comment]
`bmap::copy_nobmap` still doesn't work properly.

File: bmap/src/lib.rs [hunk 1]
```diff
@@ -79,3 +78,17 @@ where
 
     Ok(())
 }
+
+pub fn copy_nobmap<I, O>(input: &mut I, output: &mut O) -> Result<(), CopyError>
+where
+    I: Read + SeekForward,
+    O: Write + SeekForward,
```
Comment (obbardc) [inline bmap/src/lib.rs:85]: we don't need the seekforward traits for I and O?

File: bmap/src/lib.rs [hunk 2]
```diff
@@ -79,3 +78,17 @@ where
...
+{
+    let mut v = Vec::new();
+    let r = input.read_to_end(&mut v).map_err(CopyError::ReadError)?;
```
Comment (obbardc) [inline bmap/src/lib.rs:88]: you're reading all of the file into memory and then writing out again, which is super inefficient

Usually (like in the case of the original `bmap::copy`) we read small chunks into memory at a time, then write those out.

It'd be interesting to try to use https://doc.rust-lang.org/std/io/fn.copy.html here as that handles the buffering for us?

File: bmap/src/lib.rs (hunk 1 above)
Comment (Razaloc) [reply bmap/src/lib.rs:85]: I think we do and I have to use it, because as is only works for some files. I think maybe is because those have no "holes". I didn't know why that was happening.

File: bmap/src/lib.rs (hunk 2 above)
Comment (Razaloc) [reply bmap/src/lib.rs]: I just updated the code to use `st::io::copy` and without the `Seekforward` requirements for I and O. I've tested and, as before, only works with some files. I don't know why yet.

Comment (Razaloc) [reply bmap/src/lib.rs]: Another possibility is that the difference is originated in the output file not being truncated since we don't have the output size at that moment.

**Comment by Razaloc** [comment]
Now works always. I'm using `std::mem::size_of_val` to figure out the size of the decompressed data and be able to truncate the file. If it's what we need I just have to arrange the commits.

**Comment by Razaloc** [comment]
Seems to work fine without applying `ftruncate` to the output file.

File: bmap-rs/src/main.rs [hunk 3]
```diff
@@ -13,6 +13,9 @@ use std::path::{Path, PathBuf};
 
 #[derive(Parser, Debug)]
 struct Copy {
+    ///Allow copying without bmap
```
Comment (obbardc) [inline bmap-rs/src/main.rs:16]:
```suggestion
    /// Allow copying without a bmap file.
```

File: bmap-rs/src/main.rs [hunk 4]
```diff
@@ -88,31 +91,35 @@ fn setup_input(path: &Path) -> Result<Decoder> {
         _ => Ok(Decoder::new(f)),
     }
 }
-
 fn copy(c: Copy) -> Result<()> {
+    match !c.nobmap {
```
Comment (obbardc) [inline bmap-rs/src/main.rs:95]: this reads quite awkwardly; nicer way could be..

```
match c.nobmap {
    true => copy_local_input_nobmap(c),
    false => copy_local_input(c)
}
```

File: bmap-rs/src/main.rs [hunk 5]
```diff
@@ -88,31 +91,35 @@ fn setup_input(path: &Path) -> Result<Decoder> {
...
         bail!("Image file doesn't exist")
     }
 
     let bmap = find_bmap(&c.image).ok_or_else(|| anyhow!("Couldn't find bmap file"))?;
     println!("Found bmap file: {}", bmap.display());
-
     let mut b = File::open(&bmap).context("Failed to open bmap file")?;
     let mut xml = String::new();
     b.read_to_string(&mut xml)?;
-
```
Comment (obbardc) [inline bmap-rs/src/main.rs:103]: spurious change ?

File: bmap-rs/src/main.rs [hunk 6]
```diff
@@ -88,31 +91,35 @@ fn setup_input(path: &Path) -> Result<Decoder> {
...
+    }
+}
+fn copy_local_input(c: Copy) -> Result<()> {
     if !c.image.exists() {
         bail!("Image file doesn't exist")
     }
 
     let bmap = find_bmap(&c.image).ok_or_else(|| anyhow!("Couldn't find bmap file"))?;
     println!("Found bmap file: {}", bmap.display());
-
```
Comment (obbardc) [inline bmap-rs/src/main.rs:99]: spurious change ?

File: bmap-rs/src/main.rs [hunk 7]
```diff
@@ -88,31 +91,35 @@ fn setup_input(path: &Path) -> Result<Decoder> {
...
     let mut b = File::open(&bmap).context("Failed to open bmap file")?;
     let mut xml = String::new();
     b.read_to_string(&mut xml)?;
-
     let bmap = Bmap::from_xml(&xml)?;
+
```
Comment (obbardc) [inline bmap-rs/src/main.rs:111]: spurious change ?

File: bmap-rs/src/main.rs [hunk 8]
```diff
@@ -88,31 +91,35 @@ fn setup_input(path: &Path) -> Result<Decoder> {
...
     let output = std::fs::OpenOptions::new()
         .write(true)
         .create(true)
         .open(c.dest)?;
 
+    let mut input = setup_input(&c.image)?;
+
     if output.metadata()?.is_file() {
-        ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
+        ftruncate(output.as_raw_fd(), bmap.total_mapped_size() as i64)
```
Comment (obbardc) [inline bmap-rs/src/main.rs:120]: unrelated change ?

File: bmap-rs/src/main.rs [hunk 9]
```diff
@@ -88,31 +91,35 @@ fn setup_input(path: &Path) -> Result<Decoder> {
...
     b.read_to_string(&mut xml)?;
-
     let bmap = Bmap::from_xml(&xml)?;
+
     let output = std::fs::OpenOptions::new()
         .write(true)
         .create(true)
         .open(c.dest)?;
 
+    let mut input = setup_input(&c.image)?;
```
Comment (obbardc) [inline bmap-rs/src/main.rs:117]: unrelated change ?

File: bmap/src/lib.rs [hunk 10]
```diff
@@ -3,10 +3,9 @@ pub use crate::bmap::*;
 mod discarder;
 pub use crate::discarder::*;
 use sha2::{Digest, Sha256};
-use thiserror::Error;
-
 use std::io::Result as IOResult;
 use std::io::{Read, Seek, SeekFrom, Write};
+use thiserror::Error;
```
Comment (obbardc) [inline bmap/src/lib.rs:8]: unrelated change ?

**Comment by Razaloc** [comment]
Parser problem solved and commits arranged

**Review by obbardc — CHANGES_REQUESTED** [review]
I think it makes sense to squash your commits into one here, since it is one additional feature
